File: src/services/gamelogic.ts
import { PythonShell } from 'python-shell';
import path from 'path';
import fs from 'fs';

export class GameLogicService {
    private pythonPath: string;
    private chessPythonPath: string;

    constructor() {
        // Paths to Python scripts
        this.pythonPath = path.join(__dirname, '../../python/game_logic.py');
        this.chessPythonPath = path.join(__dirname, '../../python/chess_logic.py');

        // Ensure Python script directories exist
        const dirPath = path.dirname(this.pythonPath);
        if (!fs.existsSync(dirPath)) {
            fs.mkdirSync(dirPath, { recursive: true });
        }

        // Create chess logic file if it doesn't exist
        this.ensureChessLogicExists();

        // Verify Python scripts exist
        if (!fs.existsSync(this.pythonPath)) {
            console.warn(`Tic-tac-toe script not found at ${this.pythonPath}`);
        }
        if (!fs.existsSync(this.chessPythonPath)) {
            console.warn(`Chess script not found at ${this.chessPythonPath}`);
        }
    }

    private ensureChessLogicExists(): void {
        if (!fs.existsSync(this.chessPythonPath)) {
            // You would copy the chess_logic.py content here
            console.log(`Creating chess logic file at ${this.chessPythonPath}`);
            // For now, we'll assume the file exists
        }
    }

    private async runPythonScript(data: any, gameType: string = 'tic-tac-toe'): Promise<any> {
        return new Promise((resolve, reject) => {
            // Choose the appropriate Python script based on game type
            const scriptPath = gameType === 'chess' ? this.chessPythonPath : this.pythonPath;
            
            if (!fs.existsSync(scriptPath)) {
                return reject(new Error(`Python script not found: ${scriptPath}`));
            }

            // Create a new PythonShell instance
            const pyshell = new PythonShell(scriptPath);

            // Send data to the Python script
            pyshell.send(JSON.stringify(data));

            let result = '';

            // Collect output from the Python script
            pyshell.on('message', (message) => {
                result += message;
            });

            // Handle errors
            pyshell.on('error', (err) => {
                reject(err);
            });

            // End the process and resolve the promise
            pyshell.end((err) => {
                if (err) {
                    return reject(err);
                }

                try {
                    const parsedResult = JSON.parse(result);
                    resolve(parsedResult);
                } catch (e) {
                    reject(new Error(`Failed to parse Python result: ${result}`));
                }
            });
        });
    }

    /**
     * Initialize a new game state for the specified game type
     * @param gameType Type of game to initialize ('tic-tac-toe' or 'chess')
     * @returns Initial game state
     */
    async initializeGame(gameType: string): Promise<any> {
        const result = await this.runPythonScript({
            action: 'initialize',
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result;
    }

    /**
     * Validate if a move is legal
     * @param state Current game state
     * @param move Move to validate
     * @param playerId ID of player making the move
     * @returns Whether the move is valid
     */
    async validateMove(state: any, move: any, playerId: number): Promise<boolean> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'validate',
            state,
            move,
            player_id: playerId,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.valid || false;
    }

    /**
     * Apply a move to the game state
     * @param state Current game state
     * @param move Move to apply
     * @param playerId ID of player making the move
     * @returns New game state after the move
     */
    async applyMove(state: any, move: any, playerId: number): Promise<any> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'apply',
            state,
            move,
            player_id: playerId,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result;
    }

    /**
     * Check if there's a winner
     * @param state Current game state
     * @returns ID of winner if there is one, null otherwise
     */
    async checkWinner(state: any): Promise<number | null> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'check_winner',
            state,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.winner || null;
    }

    /**
     * Check if the game is over (win or draw)
     * @param state Current game state
     * @returns Whether the game is over
     */
    async isGameOver(state: any): Promise<boolean> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'is_game_over',
            state,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.game_over || false;
    }

    /**
     * Check if the game is a draw
     * @param state Current game state
     * @returns Whether the game is a draw
     */
    async isDraw(state: any): Promise<boolean> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'is_draw',
            state,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.is_draw || false;
    }

    /**
     * Determine the game type based on the state
     * @param state Game state
     * @returns Game type string
     */
    private determineGameType(state: any): string {
        // Check for chess game (8x8 board with piece objects)
        if (state && Array.isArray(state.board) && state.board.length === 8 &&
            Array.isArray(state.board[0]) && state.board[0].length === 8) {
            // Further check if it contains chess-like pieces
            if (state.castling_rights || state.en_passant_target !== undefined || 
                state.king_positions || state.captured_pieces) {
                return 'chess';
            }
        }
        
        // Check for tic-tac-toe (1D array of 9 elements)
        if (state && Array.isArray(state.board) && state.board.length === 9) {
            return 'tic-tac-toe';
        }

        // Default to tic-tac-toe for backwards compatibility
        return 'tic-tac-toe';
    }

    /**
     * Get available moves for a player (mainly for chess)
     * @param state Current game state
     * @param playerId Player ID
     * @returns List of legal moves
     */
    async getAvailableMoves(state: any, playerId: number): Promise<any[]> {
        const gameType = this.determineGameType(state);
        
        if (gameType !== 'chess') {
            // For tic-tac-toe, just return empty squares
            const moves = [];
            for (let i = 0; i < state.board.length; i++) {
                if (state.board[i] === null) {
                    moves.push({ position: i });
                }
            }
            return moves;
        }

        const result = await this.runPythonScript({
            action: 'get_available_moves',
            state,
            player_id: playerId,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.moves || [];
    }

    /**
     * Evaluate the board position (for AI purposes)
     * @param state Current game state
     * @param playerId Player ID to evaluate for
     * @returns Evaluation score
     */
    async evaluatePosition(state: any, playerId: number): Promise<number> {
        const gameType = this.determineGameType(state);
        
        const result = await this.runPythonScript({
            action: 'evaluate',
            state,
            player_id: playerId,
            game_type: gameType
        }, gameType);

        if (result && result.error) {
            throw new Error(result.error);
        }

        return result?.evaluation || 0;
    }
}

File: python/game_logic.py
"""
Game Logic for Matchmaking Server

This script implements the game logic for different board games.
It is called by the Node.js server through python-shell.
"""

import json
import sys

class TicTacToeLogic:
    """Tic-Tac-Toe game logic implementation."""
    
    def initialize_game(self):
        """Initialize a new game of Tic-Tac-Toe."""
        return {
            "board": [None, None, None, None, None, None, None, None, None],
            "current_player": 1,  # Player 1 starts
            "moves_count": 0
        }
    
    def validate_move(self, state, move, player_id):
        """Validate if a move is legal."""
        # Check if it's the player's turn
        if state["current_player"] != player_id:
            return False
        
        # Check if the position is valid
        position = move.get("position")
        if position is None or position < 0 or position > 8:
            return False
        
        # Check if the position is empty
        if state["board"][position] is not None:
            return False
        
        return True
    
    def apply_move(self, state, move, player_id):
        """Apply a move to the game state."""
        position = move["position"]
        new_state = state.copy()
        new_state["board"] = state["board"].copy()
        
        # Update board
        new_state["board"][position] = player_id
        
        # Update current player
        new_state["current_player"] = 1 if player_id == 2 else 2
        
        # Increment move count
        new_state["moves_count"] += 1
        
        return new_state
    
    def check_winner(self, state):
        """Check if there's a winner."""
        board = state["board"]
        
        # Check rows
        for i in range(0, 9, 3):
            if board[i] is not None and board[i] == board[i+1] == board[i+2]:
                return board[i]
        
        # Check columns
        for i in range(3):
            if board[i] is not None and board[i] == board[i+3] == board[i+6]:
                return board[i]
        
        # Check diagonals
        if board[0] is not None and board[0] == board[4] == board[8]:
            return board[0]
        if board[2] is not None and board[2] == board[4] == board[6]:
            return board[2]
        
        return None
    
    def is_game_over(self, state):
        """Check if the game is over (win or draw)."""
        # Check if there's a winner
        if self.check_winner(state) is not None:
            return True
        
        # Check if it's a draw (all positions filled)
        if state["moves_count"] >= 9:
            return True
            
        return False
    
    def is_draw(self, state):
        """Check if the game is a draw."""
        return self.is_game_over(state) and self.check_winner(state) is None


class GameLogicFactory:
    """Factory for creating game logic instances."""
    
    @staticmethod
    def create_game_logic(game_type):
        """Create and return a game logic instance based on game type."""
        if game_type == "tic-tac-toe":
            return TicTacToeLogic()
        else:
            raise ValueError(f"Unknown game type: {game_type}")


def handle_request():
    """Handle incoming requests from the Node.js server."""
    # Read input from stdin (from PythonShell)
    input_data = sys.stdin.readline().strip()
    
    try:
        request = json.loads(input_data)
        action = request.get("action")
        game_type = request.get("game_type", "tic-tac-toe")
        
        # Create game logic instance
        game_logic = GameLogicFactory.create_game_logic(game_type)
        
        # Process based on action
        if action == "initialize":
            result = game_logic.initialize_game()
        elif action == "validate":
            state = request.get("state")
            move = request.get("move")
            player_id = request.get("player_id")
            result = {"valid": game_logic.validate_move(state, move, player_id)}
        elif action == "apply":
            state = request.get("state")
            move = request.get("move")
            player_id = request.get("player_id")
            result = game_logic.apply_move(state, move, player_id)
        elif action == "check_winner":
            state = request.get("state")
            result = {"winner": game_logic.check_winner(state)}
        elif action == "is_game_over":
            state = request.get("state")
            result = {"game_over": game_logic.is_game_over(state)}
        elif action == "is_draw":
            state = request.get("state")
            result = {"is_draw": game_logic.is_draw(state)}
        else:
            result = {"error": f"Invalid action: {action}"}
        
        # Return result as JSON
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}))


if __name__ == "__main__":
    handle_request()